sematic/versions.py

```diff
@@ -7,8 +7,8 @@
 # Represents the version of the client, server, and all other parts of
 # the sdk. Should be bumped any time a release is made. Should be set
 # to whatever is the version after the most recent one in changelog.md,
-# as well as the version for the sematic wheel in wheel_version.bzl
-CURRENT_VERSION = (0, 23, 0)
+# as well as the version for the sematic wheel in wheel_constants.bzl
+CURRENT_VERSION = (0, 24, 0)
```

Comment:
Are we sure the client version shouldn't change?

Reply:
I'm about to do some backwards compatibility testing to verify. Will document in the PR description once I'm done

Reply:
Tested this by happenstance. The new Resolver does not work with the 0.23.0 server:

Server log:
```
2023-02-08 19:42:50,470 [INFO] gunicorn.access: 52.42.50.88 - - [08/Feb/2023:19:42:50 +0000] "GET /api/v1/meta/versions HTTP/1.1" 200 52 "-" "python-requests/2.28.1"
2023-02-08 19:42:50,486 [INFO] gunicorn.access: 52.42.50.88 - - [08/Feb/2023:19:42:50 +0000] "GET /api/v1/storage/artifacts/4bdc9789f62fda5812c5f21d5e44650e910205e1/location HTTP/1.1" 200 924 "-" "python-requests/2.28.1"
2023-02-08 19:43:00,512 [INFO] gunicorn.access: 52.42.50.88 - - [08/Feb/2023:19:43:00 +0000] "GET /api/v1/storage/artifacts/4bdc9789f62fda5812c5f21d5e44650e910205e1/location HTTP/1.1" 200 924 "-" "python-requests/2.28.1"
2023-02-08 19:43:11,529 [INFO] gunicorn.access: 52.42.50.88 - - [08/Feb/2023:19:43:11 +0000] "GET /api/v1/storage/artifacts/4bdc9789f62fda5812c5f21d5e44650e910205e1/location HTTP/1.1" 200 924 "-" "python-requests/2.28.1"
2023-02-08 19:43:11,560 [INFO] gunicorn.access: 52.42.50.88 - - [08/Feb/2023:19:43:11 +0000] "GET /api/v1/resolutions/5387506d022d4808a5a2fcad4112c92e/external_resources HTTP/1.1" 200 26 "-" "python-requests/2.28.1"
2023-02-08 19:43:11,588 [INFO] gunicorn.access: 52.42.50.88 - - [08/Feb/2023:19:43:11 +0000] "GET /api/v1/runs/5387506d022d4808a5a2fcad4112c92e HTTP/1.1" 404 63 "-" "python-requests/2.28.1"
```

Resolver log:
```
tudor@ip-10-99-0-168:~/sematic$ bazel run sematic/examples/testing_pipeline:__main__ -- --cloud --detach
INFO: Analyzed target //sematic/examples/testing_pipeline:__main__ (1 packages loaded, 1 target configured).
INFO: Found 1 target...
Target //sematic/examples/testing_pipeline:__main__ up-to-date:
  bazel-bin/sematic/examples/testing_pipeline/__main__.sh
INFO: Elapsed time: 0.447s, Critical Path: 0.01s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
INFO: Analyzed target //sematic/examples/testing_pipeline:__main___default_push (91 packages loaded, 11629 targets configured).
INFO: Found 1 target...
Target //sematic/examples/testing_pipeline:__main___default_push up-to-date:
  bazel-bin/sematic/examples/testing_pipeline/__main___default_push.digest
  bazel-bin/sematic/examples/testing_pipeline/__main___default_push
INFO: Elapsed time: 43.243s, Critical Path: 38.77s
INFO: 11 processes: 5 internal, 6 linux-sandbox.
INFO: Build completed successfully, 11 total actions
INFO: Build completed successfully, 11 total actions
2023/02/08 19:42:33 Successfully pushed Docker image to 558717131297.dkr.ecr.us-west-2.amazonaws.com/sematic-dev:__main___default - 558717131297.dkr.ecr.us-west-2.amazonaws.com/sematic-dev@sha256:823c4c166d55a2ba53589211db41a21dc58a4af3e4e1b84aa729b345c670ea69
INFO: Analyzed target //sematic/examples/testing_pipeline:__main___binary (0 packages loaded, 20 targets configured).
INFO: Found 1 target...
Target //sematic/examples/testing_pipeline:__main___binary up-to-date:
  bazel-bin/sematic/examples/testing_pipeline/__main___binary
INFO: Elapsed time: 11.212s, Critical Path: 10.44s
INFO: 4 processes: 4 internal.
INFO: Build completed successfully, 4 total actions
INFO: Build completed successfully, 4 total actions
INFO:__main__:Command line arguments: Namespace(cache_namespace=None, cloud=True, detach=True, exit_code=None, expand_shared_memory=False, external_resource=False, fan_out=0, inline=False, log_level='INFO', max_parallelism=None, nested=False, oom=False, raise_retry_probability=None, ray_resource=False, rerun_from=None, should_raise=False, silent=False, sleep_time=0, virtual_funcs=False)
INFO:__main__:Pipeline arguments: {'external_resource': False, 'ray_resource': False, 'sleep_time': 0, 'raise_retry_probability': None, 'inline': False, 'virtual_funcs': False, 'nested': False, 'oom': False, 'fan_out': 0, 'should_raise': False, 'exit_code': None}
INFO:__main__:Invoking the pipeline...
WARNING:sematic.utils.retry:Expecting value: line 1 column 1 (char 0), retrying in 10 seconds...
WARNING:sematic.utils.retry:Expecting value: line 1 column 1 (char 0), retrying in 11 seconds...
ERROR:sematic.api_client:Server returned 404 for GET https://dev.sematic.cloud/api/v1/runs/5387506d022d4808a5a2fcad4112c92e: {"error": "No runs with id '5387506d022d4808a5a2fcad4112c92e'"}
ERROR:sematic.resolvers.state_machine_resolver:Unable to fail resolution: Resource https://dev.sematic.cloud/api/v1/runs/5387506d022d4808a5a2fcad4112c92e was not found
Traceback (most recent call last):
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/pip_dependencies38_requests/site-packages/requests/models.py", line 971, in json
    return complexjson.loads(self.text, **kwargs)
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/external/python3_8_x86_64-unknown-linux-gnu/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/external/python3_8_x86_64-unknown-linux-gnu/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/external/python3_8_x86_64-unknown-linux-gnu/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/examples/testing_pipeline/__main__.py", line 328, in <module>
    main()
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/examples/testing_pipeline/__main__.py", line 323, in main
    result = future.resolve(resolver)
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/future.py", line 44, in resolve
    self.value = resolver.resolve(self)
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/resolvers/cloud_resolver.py", line 127, in resolve
    return self._detach_resolution(future)
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/resolvers/cloud_resolver.py", line 227, in _detach_resolution
    run = self._populate_run_and_artifacts(future)
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/resolvers/local_resolver.py", line 219, in _populate_run_and_artifacts
    run = self._populate_graph(future)
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/resolvers/local_resolver.py", line 689, in _populate_graph
    artifact = self._make_artifact(
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/resolvers/local_resolver.py", line 660, in _make_artifact
    api_client.store_artifact_bytes(artifact.id, payload)
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/utils/retry.py", line 30, in wrapper
    return caller(f, *args, **kwargs)
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/utils/retry.py", line 137, in retry_decorator
    return __retry_internal(
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/utils/retry.py", line 77, in __retry_internal
    return f()
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/api_client.py", line 95, in store_artifact_bytes
    response = _get(f"/storage/artifacts/{artifact_id}/location")
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/utils/retry.py", line 30, in wrapper
    return caller(f, *args, **kwargs)
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/utils/retry.py", line 137, in retry_decorator
    return __retry_internal(
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/utils/retry.py", line 77, in __retry_internal
    return f()
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/sematic/sematic/api_client.py", line 418, in _get
    return response.json()
  File "/home/tudor/.cache/bazel/_bazel_tudor/33dcc00034a66e07dc9fd2f996f93d80/execroot/sematic/bazel-out/k8-fastbuild/bin/sematic/examples/testing_pipeline/__main___binary.runfiles/pip_dependencies38_requests/site-packages/requests/models.py", line 975, in json
    raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```